Add ERC: Perpetual Contract NFTs as Collateral (#127)
ethereum/ERCs: ERCS/erc-7565.md

---
eip: 7565
title: Perpetual Contract NFTs as Collateral for Liquidity Provision
description: A standard for representing locked financial assets as Non-Fungible Tokens (NFTs), enabling these NFTs to be used as collateral for borrowing funds in Decentralized Finance (DeFi). This ERC also integrates the concept of rentable NFTs (via ERC-4907) to facilitate liquidity provision.
author: Hyoungsung Kim (@HyoungsungKim) <hyougnsung@keti.re.kr>, Yong-Suk Park <yspark@keti.re.kr>, Hyun-Sik Kim <hskim@keti.re.kr>
discussions-to: https://ethereum-magicians.org/t/erc-7565-proposal-perpetual-contract-nft-for-defi-composability/16790
status: Draft
type: Standards Track
category: ERC
created: 2023-11-27
requires: 721, 4907
---

## Abstract

> This ERC proposes a mechanism where a person (referred to as the "Asset Owner") can collateralize NFTs that represent locked deposits or assets, to borrow funds against them. These NFTs represent the right to claim the underlying assets, along with any accrued benefits, after a predefined maturity period. For an academic article, please visit [IEEE Xplore](https://ieeexplore.ieee.org/document/9967987/).

Reviewer: Please cite this external resource using the [format defined in EIP-1](https://eips.ethereum.org/EIPS/eip-1#digital-object-identifier-system). It might look something like this:

````suggestion
This ERC proposes a mechanism where a person (referred to as the "Asset Owner") can collateralize NFTs that represent locked deposits or assets, to borrow funds against them. These NFTs represent the right to claim the underlying assets, along with any accrued benefits, after a predefined maturity period. [^1]

[^1]:
    ```csl-json
    {
        "container-title": "IEEE Access",
        "author": [
            {
                "given": "Hyoungsung",
                "family": "Kim"
            },
            {
                "given": "Hyun-Sik",
                "family": "Kim"
            },
            {
                "given": "Yong-Suk",
                "family": "Park"
            }
        ],
        "DOI": "10.1109/ACCESS.2022.3225884",
        "URL": "https://ieeexplore.ieee.org/document/9967987",
        "type": "article-journal",
        "id": 9967987,
        "citation-label": "9967987",
        "issue": null,
        "issued": {
            "date-parts": [
                [
                    2022
                ]
            ]
        },
        "keyword": "Contracts;Nonfungible tokens;Cryptocurrency;Finance;Smart contracts;Blockchains;Financial services;Automated market maker (AMM);blockchain;decentralized exchange (DEX);decentralized finance (DeFi);Ethereum;liquidity pool (LP);non-fungible token (NFT);uniswap",
        "page": "126802-126814",
        "title": "Perpetual Contract NFT as Collateral for DeFi Composability",
        "volume": 10
    }
    ```
````

## Motivation

The rapidly evolving landscape of DeFi has introduced various mechanisms for asset locking, offering benefits like interest and voting rights. However, one of the significant challenges in this space is maintaining liquidity while these assets are locked. This ERC addresses this challenge by proposing a method to generate profit from locked assets using ERC-721 and ERC-4907.

> In DeFi services, such as Uniswap v3, liquidity providers contribute assets to pools and receive NFTs representing their stake. These NFTs denote the rights to the assets and the associated benefits, but they also lock the assets in the pool, often causing liquidity challenges for the providers. The current practice requires providers to withdraw their assets for urgent liquidity needs, adversely affecting the pool's liquidity and potentially increasing slippage during asset swaps.

Reviewer: We have a slight preference for not explicitly mentioning specific commercial products.

    // @param tokenId The NFT that was used as collateral
    // @param repayAmount The amount of funds to be repaid
    function repayLoan(uint256 tokenId, uint256 repayAmount) external;

    // Allows querying the loan terms for a given NFT
    // @param tokenId The NFT used as collateral
    // @return loanAmount The amount of funds borrowed
    // @return interestRate The interest rate for the loan
    // @return loanDuration The duration of the loan
    // @return loanDueDate The due date for the loan repayment
    function getLoanTerms(uint256 tokenId) external view returns (uint256 loanAmount, uint256 interestRate, uint256 loanDuration, uint256 loanDueDate);

    // Allows querying the current owner of the NFT
    // @param tokenId The NFT in question
    // @return The address of the current owner
    function currentOwner(uint256 tokenId) external view returns (address);

    // View the total amount required to repay the loan for a given NFT
    // @param tokenId The NFT used as collateral
    // @return The total amount required to repay the loan, including interest
    function viewRepayAmount(uint256 tokenId) external view returns (uint256);
}
```

> Event `Collateralized`:

Reviewer: I'd recommend making these subsections:

```suggestion
#### Event: `Collateralized`
```

This generates tags that can be used for deep linking, and makes them show up in the table of contents.

> - Implementation Suggestion: MUST be emitted when the collateralize function is successfully executed.

Reviewer: For consistency with other proposals, I'd remove the "Implementation Suggestion" prefix for all of these. Just use the RFC 2119 keywords:

```suggestion
The `Collateralized` event MUST be emitted when the `collateralize` function is successfully executed.
```

- Usage: Logs the event of a loan being repaid and the corresponding NFT being released from collateral.

Event `Defaulted`:

- Implementation Suggestion: MUST be emitted in scenarios where the loan defaults and the NFT is transferred to the lender.
- Usage: Used to log the event of a loan default and the transfer of the NFT to the lender.

Function `collateralize`:

- Implementation Suggestion: SHOULD be implemented as `external`.
- Usage: Allows an NFT owner to collateralize their NFT to receive a loan.

Function `repayLoan`:

- Implementation Suggestion: SHOULD be implemented as `external`.
- Usage: Enables an NFT owner to repay their loan and reclaim their NFT.
  
Function `getLoanTerms`:

- Implementation Suggestion: MAY be implemented as `external` `view`.
- Usage: Allows querying the loan terms for a given NFT.

Function `currentOwner`:

- Implementation Suggestion: MAY be implemented as `external` `view`.
- Usage: Enables querying the current owner of a specific NFT.

Function `viewRepayAmount`:

- Implementation Suggestion: MAY be implemented as `external` `view`.
- Usage: Enables querying the current repay amount of a specific NFT.
  

> This is what I write, but I think what you give me really reflect well what I want to propose. Please update what I write

Reviewer: ```suggestion
<!-- This is what I write, but I think what you give me really reflect well what I want to propose. Please update what I write -->
```

This seems like a note to someone in particular, not part of the actual proposal. Please use an HTML-style comment to hide it for now.

- DeFi Composability: The strategic emphasis on DeFi composability, particularly the integration between asset-locking and collateralizing services, is pivotal for this standard. This approach aims to streamline the adoption of the standard across diverse DeFi platforms and services, fostering seamless connections within the DeFi ecosystem.

### Alternate Designs and Related Work

- Comparison with ERC-4907: While ERC-4907 also introduces a dual-role model for NFTs (owner and user), our standard focuses specifically on the use of NFTs for collateralization in financial transactions, diverging from ERC-4907’s rental-oriented approach.

- Improvement Over Traditional Collateralization Methods: Compared to traditional DeFi collateralization, which often requires complete asset lock-up, this standard proposes a more dynamic and flexible model that allows for continued liquidity access.

## Backwards Compatibility

Fully compatible with ERC-721 and integrates with ERC-4907 for renting NFTs.

## Test Cases

### Test contract

```solidity

> // SPDX-License-Identifier: MIT

Reviewer: Code in the EIP itself _must_ be CC0-1.0 licensed. I permit a few other licenses (including `MIT`) in the assets directory, if you want to move this there.

        // If the borrower repays before the loan duration ends, they are charged interest only for the time the loan was held.
        // For example, if the annual interest rate is 5% and the borrower repays in half the loan term, they pay only 2.5% interest.
        uint256 elapsed = block.timestamp > (_loans[tokenId].loanStartTime + _loans[tokenId].loanDuration) 
                        ? _loans[tokenId].loanDuration  / 1 hours
                        : (block.timestamp - _loans[tokenId].loanStartTime) / 1 hours;

        // Round up
        // Example: 15/4 = 3.75
        // round((15 + 4 - 1)/4) = 4, round((15/4) = 3)
        uint256 interest = ((_loans[tokenId].loanAmount * _loans[tokenId].interestRate / 100) * elapsed + (_loans[tokenId].loanDuration / 1 hours) - 1) / 
                       (_loans[tokenId].loanDuration / 1 hours);

        // Calculate the total amount due
        uint256 totalDue = _loans[tokenId].loanAmount + interest;

        return totalDue;
    }

    // Additional functions and logic to handle loan defaults, transfers, and other aspects of the NFT lifecycle
}
```

## Security Considerations

> Needs discussion.

Reviewer: If you use an HTML-style comment, the linter will make sure you fill this in:

```suggestion
<!-- Needs discussion. -->
```

## Copyright

Copyright and related rights waived via [CC0](../LICENSE.md).

> 
> ## Citation
> 
> TBD

Reviewer: ```suggestion
```

Our build system will generate this section for you.
